src/Parser/Druid/Feral/Modules/Talents/Bloodtalons.js

```diff
@@ -0,0 +1,341 @@
+ * - Single target attacks that do initial damage consume a stack so long as that damage connects.
+ * - AoE abilities consume a charge if they succeed in hitting at least 1 enemy.
+ * - Rip (DoT with no initial damage) success is detected by watching for debuff apply or refresh.
+ */
+class Bloodtalons extends Analyzer {
+  spenders = null;
+  currentStack = 0;
+  expireTime = null;
+  lastSpender = null;
+
+  generated = 0;
+  used = 0;
+  expired = 0;
+  remainAfterFight = 0;
+  overwritten = 0;
+
+  constructor(...args) {
+    super(...args);
+
+    if (POTENTIAL_SPENDERS.length > CHART_COLOR_SPENDERS.length) {
+      throw new Error(`Bloodtalons has ${POTENTIAL_SPENDERS.length} potential spenders but only ${CHART_COLOR_SPENDERS.length} chart colors defined.`);
+    }
+    this.active = this.selectedCombatant.hasTalent(SPELLS.BLOODTALONS_TALENT.id);
+    if (this.active) {
+      /**
+       * The spenders array could be built during parsing, but doing it at the start means it'll
+       * always list the spenders in the same order. That should make comparison between fights
+       * or players easier for the user.
+       */
+      this.spenders = POTENTIAL_SPENDERS.map((spender, index) => ({
+        spell: spender,
+        chartColor: CHART_COLOR_SPENDERS[index],
+        count: 0,
+      }));
+    }
+  }
+
+  on_finished() {
+    this.remainAfterFight = this.currentStack;
+    debug && console.log(`${this.owner.formatTimestamp(this.owner.fight.end_time, 3)} fight ended with ${this.currentStack} Bloodtalons charges remaining`);
+
+    if (debug && this.generated !== (this.used + this.expired + this.remainAfterFight + this.overwritten)) {
+      console.warn(`Not all Bloodtalon charges accounted for. Generated: ${this.generated}, used: ${this.used}, expired: ${this.expired}, remainAfterFight: ${this.remainAfterFight}, overwritten: ${this.overwritten}`);
+    }
+  }
+
+  on_toPlayer_applybuff(event) {
+    if (SPELLS.BLOODTALONS_BUFF.id !== event.ability.guid) {
+      return;
+    }
+    debug && console.log(`${this.owner.formatTimestamp(event.timestamp, 3)} generated 2 Bloodtalons charges`);
+    this.generated += 2;
+    this.currentStack = 2;
+    this.expireTime = event.timestamp + BLOODTALONS_BUFF_DURATION;
+  }
+
+  on_toPlayer_refreshbuff(event) {
+    if (SPELLS.BLOODTALONS_BUFF.id !== event.ability.guid) {
+      return;
+    }
+    debug && console.log(`${this.owner.formatTimestamp(event.timestamp, 3)} generated 2 Bloodtalons charges, overwriting ${this.currentStack}`);
+    this.overwritten += this.currentStack;
+    this.generated += 2;
+    this.currentStack = 2;
+
+    // pandemic doesn't apply
+    this.expireTime = event.timestamp + BLOODTALONS_BUFF_DURATION;
+  }
+
+  on_toPlayer_removebuff(event) {
+    if (SPELLS.BLOODTALONS_BUFF.id !== event.ability.guid ||
+        this.currentStack === 0 || !this.expireTime ||
+        Math.abs(this.expireTime - event.timestamp) > EXPIRE_WINDOW) {
```

Comment:
I reckon the `Math.abs` might hide bugs with your implementation, is there any reason you don't do `event.timestamp - this.expireTime`?